### PR TITLE
Code police was here and doesn't like //@ts-ignore

### PR DESCRIPTION
**Pull request by AlCalzone**
Best read this with ignore whitespace changes on

### CODE REVIEW

File: io-package.json [hunk 1]
```diff
@@ -190,21 +190,25 @@
 		"store_quarters" : false,
 		"store_years" : false,
 		"store_period" : false,
-		"keep_days": "7",
```
Comment (AlCalzone) [inline io-package.json]: You are using those like numbers, so make them numbers!

File: io-package.json [hunk 2]
```diff
@@ -190,21 +190,25 @@
...
+		"basic_price_power" : 15,
+        "unit_price_heatpump": 0.24,
+        "unit_price_heatpump_night": 0.22,
+        "basic_price_heatpump": 31,
+		"unit_price_water" : 0.07,
+		"basic_price_water" : 22,
+        "unit_price_oil": 0.12,
+        "basic_price_oil" : 31,
+        "unit_price_gas": 0.12,
+        "basic_price_gas" : 31,
```
Comment (AlCalzone) [inline io-package.json]: I added a few missing ones, you might want to check the default values.

File: main.js [hunk 3]
```diff
@@ -16,7 +16,6 @@ const months = JSON.parse('["01_January","02_February","03_March","04_April","05
 // Create variables for object arrays
 const history    = {};
 const aliasMap   = {};
-const wh_start_val = [];
```
Comment (AlCalzone) [inline main.js]: unused

File: main.js [hunk 4]
```diff
@@ -60,7 +59,6 @@ class Sourceanalytix extends utils.Adapter {
 		this.subscribeForeignObjects("*");
 			
 		this.objects.getObjectView("custom", "state", {}, (err, doc) => {
-			let count = 0;
```
Comment (AlCalzone) [inline main.js]: only written to, never read

File: main.js [hunk 5]
```diff
@@ -255,13 +252,13 @@ class Sourceanalytix extends utils.Adapter {
 		cron.schedule("0 0 * * *", async () => {
 			// get current meter value
 			const reading = await this.getForeignStateAsync(obj_array._id);
+			if (!reading) return;
```
Comment (AlCalzone) [inline main.js]: Avoid doing something if the state is missing or has a null value

File: main.js [hunk 6]
```diff
@@ -255,13 +252,13 @@ class Sourceanalytix extends utils.Adapter {
...
 			const calc_reading = this.unit_calc_fact(obj_array, reading.val);
 
 			// Extend object with start value day
 			obj.common.custom[this.namespace].start_day = calc_reading;
 			this.log.debug("Object content custom current : " + JSON.stringify(obj));
 
-			//@ts-ignore Issue in recognized obj correctly, must be fixed in template
```
Comment (AlCalzone) [inline main.js]: Was fixed by reinstalling @iobroker/adapter-core (thus updating @types/iobroker)

File: main.js [hunk 7]
```diff
@@ -540,7 +537,7 @@ class Sourceanalytix extends utils.Adapter {
 			objekt.common = {
 				name: alias
 			};
-			this.extendObject(device, objekt, function (err) {
+			this.extendObject(device, objekt, (err) => {
```
Comment (AlCalzone) [inline main.js]: Use arrow function so `this` is defined.

File: main.js [hunk 8]
```diff
@@ -693,33 +690,35 @@ class Sourceanalytix extends utils.Adapter {
 
 			default:
 				this.log.error("Error in case handling of cost type identificaton : " + obj_cust.state_type);
-				skip_calc = true;
+				return;
```
Comment (AlCalzone) [inline main.js]: Returning instead of the if branch allows TypeScript to know that `cost_unit` is defined.

**Comment by AlCalzone** [comment]
Best read this with ignore whitespace changes on

File: main.js [hunk 9]
```diff
@@ -693,33 +691,35 @@ class Sourceanalytix extends utils.Adapter {
...
 
 		// Disable initalisatiton in case of w value is received
-		if (this.defineUnit(obj_cont) === "w"){skip_calc === true;}
+		if (this.defineUnit(obj_cont) === "w") return;
 
-		if (skip_calc === false){
+		// Get current value from meter
+		const reading = await this.getForeignStateAsync(id._id);
+		if (!reading) {
+			// maybe log something?
```
Comment (AlCalzone) [inline main.js]: Its best to stop executing the function here - not sure if you want to log something.

File: io-package.json (hunk 1 above)
Comment (DutchmanNL) [reply io-package.json]: good one thank you !

File: main.js (hunk 3 above)
Comment (DutchmanNL) [reply main.js]: not yet used, working an implementing it were this variable is needed

File: io-package.json (hunk 2 above)
Comment (DutchmanNL) [reply io-package.json]: thank you, figuring out what happened here as it looks in the repro but somehow not into the code

**Pull request merged**